apps/pkeyutl.c

```diff
@@ -371,14 +409,27 @@ int pkeyutl_main(int argc, char **argv)
         buf_outlen = kdflen;
         rv = 1;
     } else {
-        rv = do_keyop(ctx, pkey_op, NULL, (size_t *)&buf_outlen,
-                      buf_in, (size_t)buf_inlen);
+        if (rawin)
+            rv = do_raw_keyop(pkey_op, ctx, md, pkey, in, NULL, 0,
+                              NULL, (size_t *)&buf_outlen);
+        else
+            rv = do_keyop(ctx, pkey_op, NULL, (size_t *)&buf_outlen,
+                          buf_in, (size_t)buf_inlen);
     }
     if (rv > 0 && buf_outlen != 0) {
         buf_out = app_malloc(buf_outlen, "buffer output");
-        rv = do_keyop(ctx, pkey_op,
-                      buf_out, (size_t *)&buf_outlen,
-                      buf_in, (size_t)buf_inlen);
+        if (rawin) {
+            if (BIO_reset(in) != 0) {
+                BIO_printf(bio_err, "Error: BIO reset failed\n");
+                goto end;
+            }
```

Comment:
This doesn't look right. IIUC you are processing the entire file twice. Once to get the length of the signature to allocate the buffer, and then again to get the signature itself. That could be a long time if the input file is very big and isn't supposed to be the way you do things. You're only supposed to process the file once and then just call EVP_DigestSignFinal twice (the first to get the sig size and then a second time to get the sig itself).

Reply:
Yes, you are right. I didn't consider this thoroughly, will fix this. Thanks!

Reply:
I fix this by adding a `saved mctx` parameter which looks not very graceful...